include/boost/json/object_view.hpp

```diff
@@ -0,0 +1,861 @@
+        virtual std::size_t  max_size(const void * data)   const = 0;
+        virtual std::size_t  capacity(const void * data)   const = 0;
+        virtual value_view         at(const void * data, string_view key) const = 0;
+        virtual bool         contains(const void*, string_view key) const = 0;
+        virtual std::size_t     count(const void*, string_view key) const = 0;
+
+        virtual boost::optional<value_view> if_contains(const void * data, string_view key) const = 0;
+
+        virtual const void * find (const void *data, string_view key) const = 0;
+
+        virtual const void * begin(const void *data) const = 0;
+        virtual const void * end  (const void *data) const = 0;
+        virtual const void * next (const void* elem, std::ptrdiff_t change) const = 0;
+        virtual std::intptr_t distance(const void* itr1, const void * itr2) const = 0;
+        virtual key_value_view_pair dereference(const void * data, const void * elem) const = 0;
+    };
+
+    kind k_ = kind::object;     // must come first
+    const void * data_;
+    const adaptor * adaptor_;
+
+    struct empty_adaptor;
+    struct object_adaptor;
+    static const empty_adaptor empty_impl_;
+    static const object_adaptor object_impl_;
+
+    template<typename Map>
+    struct map_adaptor;
+
+    template<typename Map>
+    struct describe_adaptor;
+
+#if defined(BOOST_DESCRIBE_CXX14)
+    template<typename Class, typename Member>
+    constexpr static auto member_convertible_to_value_view_impl(Member (Class::*)) -> const std::decay_t<Member> &;
+    template<typename Member>
+    using member_convertible_to_value_view = std::is_convertible<decltype(member_convertible_to_value_view_impl(Member::pointer)), value_view>;
+#endif
+public:
+    /** The type of _Allocator_ returned by @ref get_allocator
+
+        This type is a @ref polymorphic_allocator.
+    */
+#ifdef BOOST_JSON_DOCS
+    using allocator_type = __see_below__;
+#else
+    using allocator_type = polymorphic_allocator<value>;
+#endif
+
+    /** The type of keys.
+
+        The function @ref string::max_size returns the
+        maximum allowed size of strings used as keys.
+    */
+    using key_type = string_view;
+
+    /// The type of mapped values
+    using mapped_type = value_view;
+
+    /// The element type
+    using value_type = key_value_view_pair;
+
+    /// The type used to represent unsigned integers
+    using size_type = std::size_t;
+
+    /// The type used to represent signed integers
+    using difference_type = std::ptrdiff_t;
+
+    /// A reference to an element
+    using reference = value_type&;
+
+    /// A const reference to an element
+    using const_reference = value_type const&;
+
+    /// A pointer to an element
+    using pointer = value_type*;
+
+    /// A const pointer to an element
+    using const_pointer = value_type const*;
+
+    /// A const random access iterator to an element
+    struct const_iterator
```

Comment:
you need to forward declare this here and put the declaration in json/impl/object_view.hpp